### PR TITLE
Add docs and demo

### PR DESCRIPTION
**Pull request by SomTambe**
Things added in this PR -

1. Completed README for the repo.
2. Add a demo for training a homography matrix with results.

Solves #12 completely and #18 .

### CODE REVIEW

File: src/geometry/adjoints.jl [hunk 1]
```diff
@@ -41,13 +41,22 @@ function ChainRulesCore.rrule(::Type{Interpolations.FilledExtrapolation}, itp, f
     return y, filledextra_const_pb
 end
 
-function ChainRulesCore.rrule(::Type{SVector{N, T}}, x...) where {N, T}
-    y = SVector{N,T}(x...)
-    function svector_const_pb(Δy)
+function ChainRulesCore.rrule(::Type{SArray{D, T, ND, L}}, x...) where {D, T, ND, L}
+    y = SArray{D, T, ND, L}(x...)
```
Comment (johnnychen94) [inline src/geometry/adjoints.jl:45]: Triangle dispatch is generally a bad code smell, and in this case we can perfectly avoid it as we don't really need `D, T, ND, L` here.

```suggestion
function ChainRulesCore.rrule(::Type{AT}, x...) where AT<:SArray
    y = AT(x...)
```

File: src/geometry/adjoints.jl [hunk 2]
```diff
@@ -41,13 +41,22 @@ function ChainRulesCore.rrule(::Type{Interpolations.FilledExtrapolation}, itp, f
...
+    function sarray_pb(Δy)
         Δy = map(t->eltype(x...)(t...), Δy)
         return NoTangent(), Δy
     end
-    return y, svector_const_pb
+    return y, sarray_pb
+end
+
+function ChainRulesCore.rrule(p::Type{RotMatrix{N, T, L}}, x) where {N, T, L}
```
Comment (johnnychen94) [inline src/geometry/adjoints.jl:53]: Same here, don't declare types that you don't use.

```suggestion
function ChainRulesCore.rrule(p::Type{RotMatrix}, x)
```

File: src/geometry/warp.jl [hunk 3]
```diff
@@ -32,18 +32,21 @@ function Homography{T}() where T
     return Homography{T}(SMatrix{3, 3, T, 9}(m))
 end
 
-function (h::Homography{T})(x::SVector{3,K}) where {T,K}
-    y = h.H * x
-    return SVector{2,T}(y[1:2] ./ y[end])
+function (h::Homography{K})(x::SVector{3, L}) where {L, K}
```
Comment (johnnychen94) [inline src/geometry/warp.jl:35]: `V` can be also removed from this and some other cases:

```suggestion
function (h::Homography{K})(x::SVector{3}) where {K}
```

File: test/geometry/adjoints.jl [hunk 4]
```diff
@@ -53,25 +53,51 @@ end
     end # TODO: Remove the @test_broken after tests work.
 end
 
-@testset "SVector{N, T} gradient" begin
+@testset "SVector{N, T} and SMatrix{N, N, T, L} gradient" begin
```
Comment (johnnychen94) [inline test/geometry/adjoints.jl:56]: Nobody will be interested in the types, just "StaticArrays" would be okay since this testset is already a subset of `Adjoints` group.

File: docs/src/tutorials/warp.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+## Importing the libraries
+```julia
+using DiffImages, ImageCore, ImageTransformations, FileIO
+```
+
+## Loading the images
+Let us load the images first. We will also convert them to `Float32` precision type since we do not need such high precision.
+```julia
+img = load("docs/src/assets/source.jpg") .|> RGB{Float32} 
+tgt = load("docs/src/assets/target.jpg") .|> RGB{Float32}
```
Comment (johnnychen94) [inline docs/src/tutorials/warp.md:15]: How would you ensure the codes are runnable? How would you make sure the images match the actual function output? How would you expect your docs reader to load an image that doesn't exist locally?

Try to use Documenter's `@example`/`@repl`/`@setup` for this. Or maybe just adopt Literate.jl.

File: docs/src/index.md [hunk 6]
```diff
@@ -2,12 +2,21 @@
...
 
-We hope you remain safe and healthy.
+Currently, we provide support to the following modules:
+- Warping modules inside of ImageTransformations.jl.
+    - which includes [`ImageTransformations.warp`](https://juliaimages.org/ImageTransformations.jl/stable/reference/#ImageTransformations.warp), with support for transformations from [CoordinateTransformations.jl](https://github.com/JuliaGeometry/CoordinateTransformations.jl).
+- Colorspace modules from [ImageCore.jl](https://github.com/JuliaImages/ImageCore.jl).
+
+In the future, we aim to extend support to kernels from [ImageFiltering.jl](https://github.com/JuliaImages/ImageFiltering.jl) as well as extend support to image derivatives which are not currently possible due to no implementations for different types of interpolants used in [Interpolations.jl](https://github.com/JuliaMath/Interpolations.jl).
+
+Have a look at the [examples](tutorials/warp.md) section of what you can achieve using this library.
```
Comment (johnnychen94) [inline docs/src/index.md:22]: Does this work? Or maybe `[examples](@ref some_id)` (also Documenter.jl syntax)

File: Project.toml [hunk 7]
```diff
@@ -13,6 +13,7 @@ ImageCore = "a09fc81d-aa75-5fe9-8630-4744c3626534"
 ImageTransformations = "02fcd773-0e25-5acc-982a-7f6622650795"
 Interpolations = "a98d9a8b-a2ab-59e6-89dd-64a1c18fca59"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Rotations = "6038ab10-8711-5258-84ad-4b1120ba62dc"
```
Comment (johnnychen94) [inline Project.toml:16]: Again, every dependency should declare its compatibility.

Actually, there's a way to work to make all these indirect dependencies. Say we don't add `Rotations` to Project.toml, we can still get everything via:

```julia
using ImageTransformations.Rotations
```

**Review by johnnychen94 — COMMENTED** [review]
Maintaining a long dependency list can easily be a nightmare, so at this stage we need to revisit the `deps` and try to reduce our direct dependencies. For example, can we remove `ChainRules`/`Flux`/`ColorVectorSpace`? My guess is they are not really needed by codes in `src/`.

Because `ColorVectorSpace` is already loaded by `ImageCore@0.9` so you don't need to `using ColorVectorSpace` again.

File: src/geometry/adjoints.jl (hunk 1 above)
Comment (SomTambe) [reply src/geometry/adjoints.jl:45]: I had particularly kept the types because of cases like this - 
```julia
SVector(1.0, 2.0, 3.0) 
SVector((1.0, 2.0, 3.0))
```
Both of them may look the same but the good thing with my rrule is that it the first and second methods will fallback to this sort of a thing `SVector{2, Float64}((1.0, 2.0, 3.0))` always, and therefore it will consistently apply the rrule to any sort of functor/method that encounters a `<: SArray`. 
Basically what I meant to say that keeping a generic rrule will result in many places where the other pipeline will break. I am already seeing many tests failing if I remove the types.

File: src/geometry/adjoints.jl (hunk 2 above)
Comment (SomTambe) [reply src/geometry/adjoints.jl:53]: Same thing as I commented above.

File: src/geometry/warp.jl (hunk 3 above)
Comment (SomTambe) [reply src/geometry/warp.jl:35]: Yes, this does the same thing. I'll change it as you say.

File: Project.toml (hunk 7 above)
Comment (SomTambe) [reply Project.toml:16]: Yeah I have updated all the compatibilities in the main repo, this repo was branched from the trfm branch just before I had made those changes. I have added all the compats.

File: docs/src/tutorials/warp.md (hunk 5 above)
Comment (SomTambe) [reply docs/src/tutorials/warp.md:15]: I have changed it to `@example` and `@repl`.

**Comment by SomTambe** [comment]
I have added the following things:

- Tutorial for Image Rotation transform
- Resolved the comments as suggested. 

We can now get this merged.

File: src/geometry/adjoints.jl (hunk 1 above)
Comment (johnnychen94) [reply src/geometry/adjoints.jl]: This is expected to be equivalent; obviously, the code is more fragile than I could expect. I don't have the bandwidth to figure it out so I'll just give it a pass.

File: docs/src/tutorials/rotate/rotate.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            out = image_mse(out, tgt)
+            out
+        end
+
+    out = ImageTransformations.warp(img, m(θ), axes(img), zero(eltype(img)))
+    println("Iteration: $i Loss: $(image_mse(out, tgt))")
+
+    θ = θ + η*∇θ
+end
+```
```
Comment (johnnychen94) [inline docs/src/tutorials/rotate/rotate.md:78]: Since this doesn't actually run during the building, maybe show some results like:

```text
Iteration: 1 Loss: xxx
Iteration: 2 Loss: xxx
Iteration: 3 Loss: xxx
Iteration: 4 Loss: xxx
Iteration: 5 Loss: xxx
...
Iteration: 96 Loss: xxx
Iteration: 97 Loss: xxx
Iteration: 98 Loss: xxx
Iteration: 99 Loss: xxx
Iteration: 100 Loss: xxx
```

or give a plot of the loss history.

File: docs/src/tutorials/homog/warp.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            out
+        end
+
+    out = ImageTransformations.warp(img, h, axes(img), zero(eltype(img)))
+    println("Iteration: $i Loss: $(image_mse(out, tgt))")
+
+    h = h.H - η * (∇H.H)
+    h = DiffImages.Homography(h |> SMatrix{3, 3, Float32, 9})
+end
+```
```
Comment (johnnychen94) [inline docs/src/tutorials/homog/warp.md:60]: The issue here is that even with all previous blocks defined as `@repl`, none of those codes are actually running; they only import or define some functions without actually running.

So maybe actually run one iteration, and explain what `∇H` is in this scenario, and then display the `1:num_iters` results.

Comment (SomTambe) [reply docs/src/tutorials/homog/warp.md:60]: Sure. I added the mathematical representation of ∇H, and also the outputs of the training loops in both of them

**Review by johnnychen94 — APPROVED** [review]

**Comment by SomTambe** [comment]
It looks good after Johnny's suggestions, I am merging this now.
